File: backend/src/Services/ResidentService.cs
using API.Entities;
using API.Interfaces;
using API.Types;

namespace API.Services;

public class ResidentService(IResidentRepository residentRepository) {

    public async Task<Page<Resident>> GetAll(string filter, int page, int limit) {

        Page<Resident> residents = await residentRepository.GetAll(filter, page, limit);

        return residents;

    }

    public async Task<Resident?> GetResidentById(int id){

        Resident? resident = await residentRepository.GetById(id);

        return resident;

    }

    public async Task<Resident?> CreateResident(int userId, int apartmentId, DateOnly expires, bool isOwner) {

        Resident resident = new() {
            UserId = userId,
            ApartmentId = apartmentId,
            Expires = expires,
            IsOwner = isOwner
        };

        try {
            return await residentRepository.Create(resident);
        } catch {
            return null;
        }

    }

    public async Task<Resident?> UpdateResident(int id, int? userId, int? apartmentId, DateOnly? expires, bool? isOwner) {

        Resident? resident = await residentRepository.GetById(id);

        if(resident == null) {
            return null;
        }

        resident.UserId = userId ?? resident.UserId;
        resident.ApartmentId = apartmentId ?? resident.ApartmentId;
        resident.Expires = expires ?? resident.Expires;
        resident.IsOwner = isOwner ?? resident.IsOwner;

        try {
            return await residentRepository.Update(resident);
        } catch {
            return null;
        }

    }

    public async Task<Resident?> DeleteResident(int id) {

        Resident? resident = await residentRepository.GetById(id);

        if(resident == null) {
            return null;
        }

        try {
            return await residentRepository.Delete(resident);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/resident.service.ts
import { APIError } from "@/models/api-error.models";
import { Page } from "@/models/page";
import { Resident } from "@/models/resident.model";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class ResidentService extends Service {

    static async getResidentsByApartment(apartmentId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Resident> | APIError>(`/api/resident/apartment/${apartmentId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createResident(userId : number, apartmentId : number, expires : Date, isOwner : boolean) {

            const response = await this.axios.post<Resident | APIError>("/api/resident", {
                userId,
                apartmentId,
                expires: expires.toISOString().split("T")[0],
                isOwner
            });

            return response.data;
    }

    static async updateResident(id : number, userId : optional<number>, apartmentId : optional<number>, expires : optional<Date>, isOwner : optional<boolean>) {

        const response = await this.axios.put<Resident | APIError>(`/api/resident/${id}`, {
            userId,
            apartmentId,
            expires: expires?.toISOString().split("T")[0],
            isOwner
        });

        return response.data;
    }

    static async deleteResident(id : number) {

        const response = await this.axios.delete<Resident | APIError>(`/api/resident/${id}`);

        return response.data;

    }

}